common/lib/capa/capa/responsetypes.py

```diff
@@ -831,16 +841,110 @@ def assign_choice_names(self):
+        elif halves_error_count == 1 and len(all_choices) > 2:
+            return_grade = round(self.get_max_score() / 2.0, 2)
+            return CorrectMap(self.answer_id, correctness='partially-correct', npoints=return_grade)
+        elif halves_error_count == 2 and len(all_choices) > 4:
+            return_grade = round(self.get_max_score() / 4.0, 2)
+            return CorrectMap(self.answer_id, correctness='partially-correct', npoints=return_grade)
+        else:
+            return CorrectMap(self.answer_id, 'incorrect')
+
     def get_score(self, student_answers):
 
+        # Setting up answer lists:
+        #  all_choices: the full list of checkboxes
+        #  student_answer: what the student actually chose (note no "s")
+        #  student_non_answer: what they didn't choose
+        #  self.correct_choices: boxes that should be checked
+        #  self.incorrect_choices: boxes that should NOT be checked
+
+        all_choices = self.correct_choices.union(self.incorrect_choices)
+
         student_answer = student_answers.get(self.answer_id, [])
 
         if not isinstance(student_answer, list):
             student_answer = [student_answer]
 
+        # "None apply" should really be a valid choice for "check all that apply",
+        # but it throws an error if all the checks are blank.
         no_empty_answer = student_answer != []
+
         student_answer = set(student_answer)
 
+        student_non_answers = all_choices - student_answer
+
+        if not no_empty_answer:
+            return CorrectMap(self.answer_id, 'incorrect')
+
+        # This below checks to see whether we're using an alternate grading scheme.
+        #  Set partial_credit="false" (or remove it) to require an exact answer for any credit.
+        #  Set partial_credit="EDC" to count each choice for equal points (Every Decision Counts).
+        #  Set partial_credit="halves" to take half credit off for each error.
+
+        tree = self.xml
+        problem_xml = tree.xpath('.')
+
+        # Partial credit type - only one type at a time right now.
+        credit_type = problem_xml[0].get('partial_credit', default=False)
+
+        try:
+            credit_type = str(credit_type).lower().strip()
+        except ValueError:
+            _ = self.capa_system.i18n.ugettext
+            # Translators: 'partial_credit' is an attribute name and should not be translated.
+            # 'EDC' and 'halves' and 'false' should also not be translated.
+            msg = _("partial_credit value should be one of 'EDC', 'halves', or 'false'.")
+            raise LoncapaProblemError(msg)
+
+        if credit_type == 'false':
+            pass
+        elif credit_type == 'halves':
+            return self.grade_via_halves(all_choices, student_answer, student_non_answers)
+        elif credit_type == 'edc':
+            return self.grade_via_edc(all_choices, student_answer, student_non_answers)
+
```

Comment:
The try/except structure works differently than what I found on a quick skim of the original code, unless I missed something. In most cases, those constructs casts to integers, booleans, and similar, which do catch invalid inputs. Casting to string, in contrast, does not catch invalid input. "Bob" will cast to string just fine, but will be an incorrect value. "Bob" will raise an error casting to an integer. 

> Are you recommending breaking the default (non-partial) credit scheme into its own function as well?

Absolutely. This is mixing together routing code and grading code in the same function, which is a bit messy. This way:

* There is a set of three graders. Each of those is simple, self-contained, performs a single function, and is well understood. 
* There is a map between grading functions and XML attributes. This has no repetition. If you add a new grading function, error messages, parsers, routers, etc. all update with just the one change. Ideally, this can be populated directly or through decorators. 
* There is a routing function. This is again, simple, single-purpose, and easily understood. Ideally, this belongs in ReponseTypes rather than each individual type.